File: src/harness_core/user_style_mutation.py
"""User-style mutation evaluation helpers.

Provides schema validation, fixture loading, and admission grouping for
user-style mutation cases that exercise how the harness handles different
input expression styles (formal, chat, terse).
"""

from __future__ import annotations

import json
from dataclasses import dataclass, asdict
from pathlib import Path
from typing import Any, Dict, List, Optional, Sequence, Tuple


# ---------------------------------------------------------------------------
# Schema constants
# ---------------------------------------------------------------------------

SCHEMA_VERSION = "user_style_mutation.v1"

VARIANT_TYPES: Tuple[str, ...] = (
    "formal_issue",
    "user_style_chat_request",
    "terse_ticket",
)

ADMISSION_OUTCOMES: Tuple[str, ...] = (
    "admitted",
    "diagnostic",
    "needs_clarification",
    "rejected",
)

CONTAMINATION_RISKS: Tuple[str, ...] = (
    "low",
    "medium",
    "high",
    "unknown",
)

ADMISSION_SCOPES: Tuple[str, ...] = (
    "admitted",
    "diagnostic",
)

REQUIRED_MUTATION_FIELDS: Sequence[str] = (
    "case_id",
    "base_fixture_id",
    "variant_type",
    "user_prompt",
    "expected_task_family",
    "expected_required_fields",
    "expected_missing_fields",
    "admission_expectation",
    "evidence_refs",
    "fixture_metadata",
)

REQUIRED_METADATA_FIELDS: Sequence[str] = (
    "fixture_id",
    "source_type",
    "freshness",
    "estimated_human_minutes",
    "difficulty",
    "contamination_risk",
    "admission_scope",
)

SOURCE_TYPES: Tuple[str, ...] = (
    "synthetic",
    "copied_real_read_only",
    "mutated_user_style",
)


# ---------------------------------------------------------------------------
# Data classes
# ---------------------------------------------------------------------------

@dataclass
class FixtureMetadata:
    """Metadata for a mutation fixture."""

    fixture_id: str
    source_type: str
    freshness: str
    estimated_human_minutes: float
    difficulty: str
    contamination_risk: str
    admission_scope: str

    def to_dict(self) -> Dict[str, Any]:
        return asdict(self)


@dataclass
class MutationCase:
    """A single user-style mutation case."""

    case_id: str
    base_fixture_id: str
    variant_type: str
    user_prompt: str
    expected_task_family: str
    expected_required_fields: List[str]
    expected_missing_fields: List[str]
    admission_expectation: str
    evidence_refs: List[str]
    fixture_metadata: FixtureMetadata
    schema_version: str = SCHEMA_VERSION

    def to_dict(self) -> Dict[str, Any]:
        d = asdict(self)
        return d

    def to_json(self, **kw: Any) -> str:
        return json.dumps(self.to_dict(), **kw)


# ---------------------------------------------------------------------------
# Validation
# ---------------------------------------------------------------------------

def validate_fixture_metadata(data: Dict[str, Any]) -> List[str]:
    """Validate a fixture_metadata dict. Returns list of violations."""
    violations: List[str] = []
    for f in REQUIRED_METADATA_FIELDS:
        if f not in data:
            violations.append(f"fixture_metadata missing required field: {f}")
    if violations:
        return violations

    if data["source_type"] not in SOURCE_TYPES:
        violations.append(
            f"source_type {data['source_type']!r} not in {SOURCE_TYPES}"
        )
    if data["contamination_risk"] not in CONTAMINATION_RISKS:
        violations.append(
            f"contamination_risk {data['contamination_risk']!r} not in {CONTAMINATION_RISKS}"
        )
    if data["admission_scope"] not in ADMISSION_SCOPES:
        violations.append(
            f"admission_scope {data['admission_scope']!r} not in {ADMISSION_SCOPES}"
        )
    if not isinstance(data["estimated_human_minutes"], (int, float)):
        violations.append("estimated_human_minutes must be numeric")
    return violations


def validate_mutation_case(data: Dict[str, Any]) -> List[str]:
    """Validate a mutation case dict against the schema. Returns violations."""
    violations: List[str] = []

    for f in REQUIRED_MUTATION_FIELDS:
        if f not in data:
            violations.append(f"missing required field: {f}")
    if violations:
        return violations

    # schema_version
    if data["schema_version"] != SCHEMA_VERSION:
        violations.append(
            f"schema_version must be {SCHEMA_VERSION}, got {data['schema_version']!r}"
        )

    # variant_type
    if data["variant_type"] not in VARIANT_TYPES:
        violations.append(
            f"variant_type {data['variant_type']!r} not in {VARIANT_TYPES}"
        )

    # admission_expectation
    if data["admission_expectation"] not in ADMISSION_OUTCOMES:
        violations.append(
            f"admission_expectation {data['admission_expectation']!r} not in {ADMISSION_OUTCOMES}"
        )

    # evidence_refs must be a list
    if not isinstance(data["evidence_refs"], list):
        violations.append("evidence_refs must be a list")

    # expected_required_fields must be a list
    if not isinstance(data["expected_required_fields"], list):
        violations.append("expected_required_fields must be a list")

    # expected_missing_fields must be a list
    if not isinstance(data["expected_missing_fields"], list):
        violations.append("expected_missing_fields must be a list")

    # fixture_metadata sub-validation
    if isinstance(data.get("fixture_metadata"), dict):
        violations.extend(validate_fixture_metadata(data["fixture_metadata"]))
    else:
        violations.append("fixture_metadata must be a dict")

    return violations


def create_mutation_case(
    *,
    case_id: str,
    base_fixture_id: str,
    variant_type: str,
    user_prompt: str,
    expected_task_family: str,
    expected_required_fields: Optional[List[str]] = None,
    expected_missing_fields: Optional[List[str]] = None,
    admission_expectation: str,
    evidence_refs: Optional[List[str]] = None,
    fixture_metadata: Optional[Dict[str, Any]] = None,
) -> MutationCase:
    """Factory for MutationCase with defaults."""
    meta = fixture_metadata or {}
    return MutationCase(
        case_id=case_id,
        base_fixture_id=base_fixture_id,
        variant_type=variant_type,
        user_prompt=user_prompt,
        expected_task_family=expected_task_family,
        expected_required_fields=expected_required_fields or [],
        expected_missing_fields=expected_missing_fields or [],
        admission_expectation=admission_expectation,
        evidence_refs=evidence_refs or [],
        fixture_metadata=FixtureMetadata(
            fixture_id=meta.get("fixture_id", case_id),
            source_type=meta.get("source_type", "mutated_user_style"),
            freshness=meta.get("freshness", "2026-05-19"),
            estimated_human_minutes=meta.get("estimated_human_minutes", 0.0),
            difficulty=meta.get("difficulty", "unknown"),
            contamination_risk=meta.get("contamination_risk", "low"),
            admission_scope=meta.get("admission_scope", "admitted"),
        ),
    )


# ---------------------------------------------------------------------------
# Fixture loading
# ---------------------------------------------------------------------------

def load_fixture(path: Path) -> Dict[str, Any]:
    """Load a single JSON fixture file."""
    with open(path, "r") as f:
        return json.load(f)


def load_and_validate_fixture(path: Path) -> Tuple[Dict[str, Any], List[str]]:
    """Load a mutation fixture and return (data, violations)."""
    data = load_fixture(path)
    return data, validate_mutation_case(data)


def load_all_fixtures(fixture_dir: Path) -> List[Tuple[str, Dict[str, Any], List[str]]]:
    """Load and validate every .json fixture in a directory.

    Returns list of (filename, data, violations) tuples.
    """
    results = []
    for p in sorted(fixture_dir.glob("*.json")):
        data = load_fixture(p)
        violations = validate_mutation_case(data)
        results.append((p.name, data, violations))
    return results


# ---------------------------------------------------------------------------
# Admission grouping
# ---------------------------------------------------------------------------

def group_by_admission(
    fixtures: List[Tuple[str, Dict[str, Any], List[str]]],
) -> Dict[str, List[Dict[str, Any]]]:
    """Group valid fixtures by their admission_expectation.

    Only includes fixtures with no validation violations.
    """
    groups: Dict[str, List[Dict[str, Any]]] = {
        outcome: [] for outcome in ADMISSION_OUTCOMES
    }
    for _filename, data, violations in fixtures:
        if not violations:
            outcome = data["admission_expectation"]
            groups[outcome].append(data)
    return groups


def group_by_variant(
    fixtures: List[Tuple[str, Dict[str, Any], List[str]]],
) -> Dict[str, List[Dict[str, Any]]]:
    """Group valid fixtures by variant_type."""
    groups: Dict[str, List[Dict[str, Any]]] = {
        vt: [] for vt in VARIANT_TYPES
    }
    for _filename, data, violations in fixtures:
        if not violations:
            vt = data["variant_type"]
            groups[vt].append(data)
    return groups


def group_by_base_fixture(
    fixtures: List[Tuple[str, Dict[str, Any], List[str]]],
) -> Dict[str, List[Dict[str, Any]]]:
    """Group valid fixtures by base_fixture_id."""
    groups: Dict[str, List[Dict[str, Any]]] = {}
    for _filename, data, violations in fixtures:
        if not violations:
            base = data["base_fixture_id"]
            groups.setdefault(base, []).append(data)
    return groups

File: engine/src/app_layer/user_style_mutation.rs
use std::collections::HashMap;

use serde::{Deserialize, Serialize};
use serde_json::{json, Value};

// ---------------------------------------------------------------------------
// Schema constants
// ---------------------------------------------------------------------------

pub const SCHEMA_VERSION: &str = "user_style_mutation.v1";

pub const VARIANT_TYPES: &[&str] = &["formal_issue", "user_style_chat_request", "terse_ticket"];

pub const ADMISSION_OUTCOMES: &[&str] =
    &["admitted", "diagnostic", "needs_clarification", "rejected"];

pub const CONTAMINATION_RISKS: &[&str] = &["low", "medium", "high", "unknown"];

pub const ADMISSION_SCOPES: &[&str] = &["admitted", "diagnostic"];

pub const SOURCE_TYPES: &[&str] = &["synthetic", "copied_real_read_only", "mutated_user_style"];

pub const REQUIRED_MUTATION_FIELDS: &[&str] = &[
    "case_id",
    "base_fixture_id",
    "variant_type",
    "user_prompt",
    "expected_task_family",
    "expected_required_fields",
    "expected_missing_fields",
    "admission_expectation",
    "evidence_refs",
    "fixture_metadata",
];

pub const REQUIRED_METADATA_FIELDS: &[&str] = &[
    "fixture_id",
    "source_type",
    "freshness",
    "estimated_human_minutes",
    "difficulty",
    "contamination_risk",
    "admission_scope",
];

// ---------------------------------------------------------------------------
// Data structs
// ---------------------------------------------------------------------------

#[derive(Clone, Debug, Serialize, Deserialize, PartialEq)]
pub struct FixtureMetadata {
    pub fixture_id: String,
    pub source_type: String,
    pub freshness: String,
    pub estimated_human_minutes: f64,
    pub difficulty: String,
    pub contamination_risk: String,
    pub admission_scope: String,
}

impl Default for FixtureMetadata {
    fn default() -> Self {
        Self {
            fixture_id: String::new(),
            source_type: "mutated_user_style".to_string(),
            freshness: "2026-05-19".to_string(),
            estimated_human_minutes: 0.0,
            difficulty: "unknown".to_string(),
            contamination_risk: "low".to_string(),
            admission_scope: "admitted".to_string(),
        }
    }
}

#[derive(Clone, Debug, Serialize, Deserialize, PartialEq)]
pub struct MutationCase {
    pub case_id: String,
    pub base_fixture_id: String,
    pub variant_type: String,
    pub user_prompt: String,
    pub expected_task_family: String,
    pub expected_required_fields: Vec<String>,
    pub expected_missing_fields: Vec<String>,
    pub admission_expectation: String,
    pub evidence_refs: Vec<String>,
    pub fixture_metadata: FixtureMetadata,
    pub schema_version: String,
}

impl Default for MutationCase {
    fn default() -> Self {
        Self {
            case_id: String::new(),
            base_fixture_id: String::new(),
            variant_type: String::new(),
            user_prompt: String::new(),
            expected_task_family: String::new(),
            expected_required_fields: Vec::new(),
            expected_missing_fields: Vec::new(),
            admission_expectation: String::new(),
            evidence_refs: Vec::new(),
            fixture_metadata: FixtureMetadata::default(),
            schema_version: SCHEMA_VERSION.to_string(),
        }
    }
}

impl MutationCase {
    pub fn to_value(&self) -> Value {
        serde_json::to_value(self).expect("MutationCase should serialize to JSON")
    }
}

// ---------------------------------------------------------------------------
// Validation
// ---------------------------------------------------------------------------

pub fn validate_fixture_metadata(data: &Value) -> Vec<String> {
    let mut violations = Vec::new();

    for &f in REQUIRED_METADATA_FIELDS {
        if data.get(f).is_none() {
            violations.push(format!("fixture_metadata missing required field: {}", f));
        }
    }
    if !violations.is_empty() {
        return violations;
    }

    if let Some(st) = data["source_type"].as_str() {
        if !SOURCE_TYPES.contains(&st) {
            violations.push(format!("source_type {:?} not in {:?}", st, SOURCE_TYPES));
        }
    }
    if let Some(cr) = data["contamination_risk"].as_str() {
        if !CONTAMINATION_RISKS.contains(&cr) {
            violations.push(format!(
                "contamination_risk {:?} not in {:?}",
                cr, CONTAMINATION_RISKS
            ));
        }
    }
    if let Some(as_val) = data["admission_scope"].as_str() {
        if !ADMISSION_SCOPES.contains(&as_val) {
            violations.push(format!(
                "admission_scope {:?} not in {:?}",
                as_val, ADMISSION_SCOPES
            ));
        }
    }
    let ehm = &data["estimated_human_minutes"];
    if !ehm.is_number() {
        violations.push("estimated_human_minutes must be numeric".to_string());
    }

    violations
}

pub fn validate_mutation_case(data: &Value) -> Vec<String> {
    let mut violations = Vec::new();

    for &f in REQUIRED_MUTATION_FIELDS {
        if data.get(f).is_none() {
            violations.push(format!("missing required field: {}", f));
        }
    }
    if !violations.is_empty() {
        return violations;
    }

    // schema_version
    if data["schema_version"].as_str() != Some(SCHEMA_VERSION) {
        violations.push(format!(
            "schema_version must be {}, got {:?}",
            SCHEMA_VERSION,
            data["schema_version"].as_str()
        ));
    }

    // variant_type
    if let Some(vt) = data["variant_type"].as_str() {
        if !VARIANT_TYPES.contains(&vt) {
            violations.push(format!("variant_type {:?} not in {:?}", vt, VARIANT_TYPES));
        }
    }

    // admission_expectation
    if let Some(ae) = data["admission_expectation"].as_str() {
        if !ADMISSION_OUTCOMES.contains(&ae) {
            violations.push(format!(
                "admission_expectation {:?} not in {:?}",
                ae, ADMISSION_OUTCOMES
            ));
        }
    }

    // evidence_refs must be a list
    if data.get("evidence_refs").is_some() && !data["evidence_refs"].is_array() {
        violations.push("evidence_refs must be a list".to_string());
    }

    // expected_required_fields must be a list
    if data.get("expected_required_fields").is_some()
        && !data["expected_required_fields"].is_array()
    {
        violations.push("expected_required_fields must be a list".to_string());
    }

    // expected_missing_fields must be a list
    if data.get("expected_missing_fields").is_some() && !data["expected_missing_fields"].is_array()
    {
        violations.push("expected_missing_fields must be a list".to_string());
    }

    // fixture_metadata sub-validation
    match data.get("fixture_metadata") {
        Some(Value::Object(_)) => {
            violations.extend(validate_fixture_metadata(&data["fixture_metadata"]));
        }
        _ => violations.push("fixture_metadata must be a dict".to_string()),
    }

    violations
}

#[allow(clippy::too_many_arguments)]
pub fn create_mutation_case(
    case_id: &str,
    base_fixture_id: &str,
    variant_type: &str,
    user_prompt: &str,
    expected_task_family: &str,
    expected_required_fields: Vec<String>,
    expected_missing_fields: Vec<String>,
    admission_expectation: &str,
    evidence_refs: Vec<String>,
    meta: Option<&Value>,
) -> MutationCase {
    let meta = meta.cloned().unwrap_or_else(|| json!({}));
    let get_str = |key: &str, default: &str| -> String {
        meta.get(key)
            .and_then(|v| v.as_str())
            .unwrap_or(default)
            .to_string()
    };
    let get_f64 = |key: &str, default: f64| -> f64 {
        meta.get(key).and_then(|v| v.as_f64()).unwrap_or(default)
    };

    MutationCase {
        case_id: case_id.to_string(),
        base_fixture_id: base_fixture_id.to_string(),
        variant_type: variant_type.to_string(),
        user_prompt: user_prompt.to_string(),
        expected_task_family: expected_task_family.to_string(),
        expected_required_fields,
        expected_missing_fields,
        admission_expectation: admission_expectation.to_string(),
        evidence_refs,
        fixture_metadata: FixtureMetadata {
            fixture_id: get_str("fixture_id", case_id),
            source_type: get_str("source_type", "mutated_user_style"),
            freshness: get_str("freshness", "2026-05-19"),
            estimated_human_minutes: get_f64("estimated_human_minutes", 0.0),
            difficulty: get_str("difficulty", "unknown"),
            contamination_risk: get_str("contamination_risk", "low"),
            admission_scope: get_str("admission_scope", "admitted"),
        },
        schema_version: SCHEMA_VERSION.to_string(),
    }
}

// ---------------------------------------------------------------------------
// Admission grouping
// ---------------------------------------------------------------------------

pub type FixtureEntry = (String, Value, Vec<String>);

pub fn group_by_admission(fixtures: &[FixtureEntry]) -> HashMap<String, Vec<Value>> {
    let mut groups: HashMap<String, Vec<Value>> = HashMap::new();
    for outcome in ADMISSION_OUTCOMES {
        groups.insert(outcome.to_string(), Vec::new());
    }
    for (_filename, data, violations) in fixtures {
        if violations.is_empty() {
            if let Some(outcome) = data.get("admission_expectation").and_then(|v| v.as_str()) {
                groups
                    .entry(outcome.to_string())
                    .or_default()
                    .push(data.clone());
            }
        }
    }
    groups
}

pub fn group_by_variant(fixtures: &[FixtureEntry]) -> HashMap<String, Vec<Value>> {
    let mut groups: HashMap<String, Vec<Value>> = HashMap::new();
    for vt in VARIANT_TYPES {
        groups.insert(vt.to_string(), Vec::new());
    }
    for (_filename, data, violations) in fixtures {
        if violations.is_empty() {
            if let Some(vt) = data.get("variant_type").and_then(|v| v.as_str()) {
                groups.entry(vt.to_string()).or_default().push(data.clone());
            }
        }
    }
    groups
}

pub fn group_by_base_fixture(fixtures: &[FixtureEntry]) -> HashMap<String, Vec<Value>> {
    let mut groups: HashMap<String, Vec<Value>> = HashMap::new();
    for (_filename, data, violations) in fixtures {
        if violations.is_empty() {
            if let Some(base) = data.get("base_fixture_id").and_then(|v| v.as_str()) {
                groups
                    .entry(base.to_string())
                    .or_default()
                    .push(data.clone());
            }
        }
    }
    groups
}

// ---------------------------------------------------------------------------
// Tests
// ---------------------------------------------------------------------------

#[cfg(test)]
mod tests {
    use super::*;
    use serde_json::json;

    fn valid_metadata() -> Value {
        json!({
            "fixture_id": "f001",
            "source_type": "mutated_user_style",
            "freshness": "2026-05-19",
            "estimated_human_minutes": 5.0,
            "difficulty": "medium",
            "contamination_risk": "low",
            "admission_scope": "admitted",
        })
    }

    fn valid_mutation_case() -> Value {
        json!({
            "schema_version": SCHEMA_VERSION,
            "case_id": "mc001",
            "base_fixture_id": "bf001",
            "variant_type": "formal_issue",
            "user_prompt": "Please fix the bug in module X",
            "expected_task_family": "bug_fix",
            "expected_required_fields": ["file_path"],
            "expected_missing_fields": [],
            "admission_expectation": "admitted",
            "evidence_refs": ["e1"],
            "fixture_metadata": valid_metadata(),
        })
    }

    #[test]
    fn test_validate_valid_metadata() {
        let v = valid_metadata();
        let violations = validate_fixture_metadata(&v);
        assert!(violations.is_empty(), "violations: {:?}", violations);
    }

    #[test]
    fn test_validate_metadata_bad_source_type() {
        let mut v = valid_metadata();
        v["source_type"] = json!("invalid_source");
        let violations = validate_fixture_metadata(&v);
        assert!(violations.iter().any(|v| v.contains("source_type")));
    }

    #[test]
    fn test_validate_metadata_bad_contamination_risk() {
        let mut v = valid_metadata();
        v["contamination_risk"] = json!("extreme");
        let violations = validate_fixture_metadata(&v);
        assert!(violations.iter().any(|v| v.contains("contamination_risk")));
    }

    #[test]
    fn test_validate_metadata_non_numeric_minutes() {
        let mut v = valid_metadata();
        v["estimated_human_minutes"] = json!("not_a_number");
        let violations = validate_fixture_metadata(&v);
        assert!(violations
            .iter()
            .any(|v| v.contains("estimated_human_minutes")));
    }

    #[test]
    fn test_validate_valid_mutation_case() {
        let v = valid_mutation_case();
        let violations = validate_mutation_case(&v);
        assert!(violations.is_empty(), "violations: {:?}", violations);
    }

    #[test]
    fn test_validate_mutation_case_bad_variant_type() {
        let mut v = valid_mutation_case();
        v["variant_type"] = json!("invalid_variant");
        let violations = validate_mutation_case(&v);
        assert!(violations.iter().any(|v| v.contains("variant_type")));
    }

    #[test]
    fn test_validate_mutation_case_bad_admission_expectation() {
        let mut v = valid_mutation_case();
        v["admission_expectation"] = json!("invalid_outcome");
        let violations = validate_mutation_case(&v);
        assert!(violations
            .iter()
            .any(|v| v.contains("admission_expectation")));
    }

    #[test]
    fn test_validate_mutation_case_missing_field() {
        let mut v = valid_mutation_case();
        v.as_object_mut().unwrap().remove("case_id");
        let violations = validate_mutation_case(&v);
        assert!(violations.iter().any(|v| v.contains("case_id")));
    }

    #[test]
    fn test_create_mutation_case_defaults() {
        let mc = create_mutation_case(
            "mc1",
            "bf1",
            "formal_issue",
            "fix bug",
            "bug_fix",
            vec![],
            vec![],
            "admitted",
            vec![],
            None,
        );
        assert_eq!(mc.case_id, "mc1");
        assert_eq!(mc.fixture_metadata.source_type, "mutated_user_style");
        assert_eq!(mc.fixture_metadata.contamination_risk, "low");
        assert_eq!(mc.schema_version, SCHEMA_VERSION);
    }

    #[test]
    fn test_create_mutation_case_with_meta() {
        let meta = json!({
            "source_type": "synthetic",
            "contamination_risk": "high",
            "estimated_human_minutes": 10.0,
        });
        let mc = create_mutation_case(
            "mc2",
            "bf2",
            "terse_ticket",
            "fix it",
            "bug_fix",
            vec!["field_a".to_string()],
            vec!["field_b".to_string()],
            "diagnostic",
            vec!["e1".to_string()],
            Some(&meta),
        );
        assert_eq!(mc.fixture_metadata.source_type, "synthetic");
        assert_eq!(mc.fixture_metadata.contamination_risk, "high");
        assert_eq!(mc.fixture_metadata.estimated_human_minutes, 10.0);
    }

    #[test]
    fn test_group_by_admission() {
        let fixtures: Vec<FixtureEntry> = vec![
            (
                "f1.json".to_string(),
                json!({"admission_expectation": "admitted", "variant_type": "formal_issue", "base_fixture_id": "bf1"}),
                vec![],
            ),
            (
                "f2.json".to_string(),
                json!({"admission_expectation": "diagnostic", "variant_type": "terse_ticket", "base_fixture_id": "bf2"}),
                vec![],
            ),
            (
                "f3.json".to_string(),
                json!({"admission_expectation": "admitted", "variant_type": "formal_issue", "base_fixture_id": "bf1"}),
                vec!["error".to_string()],
            ),
        ];
        let groups = group_by_admission(&fixtures);
        assert_eq!(groups["admitted"].len(), 1);
        assert_eq!(groups["diagnostic"].len(), 1);
        assert_eq!(groups["rejected"].len(), 0);
    }

    #[test]
    fn test_group_by_variant() {
        let fixtures: Vec<FixtureEntry> = vec![
            (
                "f1.json".to_string(),
                json!({"admission_expectation": "admitted", "variant_type": "formal_issue", "base_fixture_id": "bf1"}),
                vec![],
            ),
            (
                "f2.json".to_string(),
                json!({"admission_expectation": "admitted", "variant_type": "formal_issue", "base_fixture_id": "bf2"}),
                vec![],
            ),
            (
                "f3.json".to_string(),
                json!({"admission_expectation": "admitted", "variant_type": "terse_ticket", "base_fixture_id": "bf3"}),
                vec![],
            ),
        ];
        let groups = group_by_variant(&fixtures);
        assert_eq!(groups["formal_issue"].len(), 2);
        assert_eq!(groups["terse_ticket"].len(), 1);
        assert_eq!(groups["user_style_chat_request"].len(), 0);
    }

    #[test]
    fn test_group_by_base_fixture() {
        let fixtures: Vec<FixtureEntry> = vec![
            (
                "f1.json".to_string(),
                json!({"admission_expectation": "admitted", "variant_type": "formal_issue", "base_fixture_id": "bf1"}),
                vec![],
            ),
            (
                "f2.json".to_string(),
                json!({"admission_expectation": "admitted", "variant_type": "terse_ticket", "base_fixture_id": "bf1"}),
                vec![],
            ),
            (
                "f3.json".to_string(),
                json!({"admission_expectation": "admitted", "variant_type": "formal_issue", "base_fixture_id": "bf2"}),
                vec![],
            ),
        ];
        let groups = group_by_base_fixture(&fixtures);
        assert_eq!(groups["bf1"].len(), 2);
        assert_eq!(groups["bf2"].len(), 1);
    }

    #[test]
    fn test_mutation_case_struct_default() {
        let mc = MutationCase::default();
        assert_eq!(mc.schema_version, SCHEMA_VERSION);
        assert!(mc.case_id.is_empty());
    }

    #[test]
    fn test_fixture_metadata_struct_default() {
        let fm = FixtureMetadata::default();
        assert_eq!(fm.source_type, "mutated_user_style");
        assert_eq!(fm.contamination_risk, "low");
        assert_eq!(fm.admission_scope, "admitted");
    }
}
